utils/multi_machine_generator/src/graph.rs

```diff
@@ -1,20 +1,45 @@
-use std::net::SocketAddr;
+use std::fmt::{Display, Formatter};
 
 use petgraph::{graph::NodeIndex, Direction, Graph};
+use serde::Serialize;
 
 /// A node of the network
-#[derive(Debug, Clone, Default)]
-pub struct MultiMachineNode {}
+#[derive(Debug, Clone)]
+pub struct MultiMachineNode {
+    addr: String,
+}
+
+/// The final configuration of a node on the network
+#[derive(Debug, Clone, Serialize)]
+pub struct MultiMachineNodeConfig {
+    addr: String,
+    parent: Option<String>,
+    port: u16,
+}
 
 /// The tree
 pub struct MultiMachineTree {
-    pub graph: Graph<MultiMachineNode, ()>,
+    pub graph: Graph<MultiMachineNode, MultiMachineEdge>,
+}
+
+pub struct MultiMachineEdge;
+
+impl Display for MultiMachineEdge {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(f, "")
+    }
+}
+
+impl Display for MultiMachineNode {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.addr)
+    }
 }
 
 impl MultiMachineNode {
     #[must_use]
-    pub fn new() -> Self {
-        Self {}
+    pub fn new(addr: String) -> Self {
```

Comment:
string seems odd, `where A: ToSocketAddrs` or something like that?

Reply:
it's because SocketAddr cannot store domain names, which we need to support.
i guess there is a type somewhere to store both cleanly, but i kept string for now for simplicity

Reply:
kinda makes sense

Reply:
Maybe check in `new` if it's legal?